### PR TITLE
abseil: force gcc 7 usage

### PR DESCRIPTION
**Pull request by iMichka**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [ ] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

### CODE REVIEW

**Comment by iMichka** [comment]
@sjackman Let's discuss the fails_with situation here, now that the gcc PR is merged. Abseil is a good small example to be able to test what is going on.

I have almost all gcc versions installed on my test system. This might be the case for any of our users. This will also often be the case in CI with mixed dependency trees.

The first thing I noticed is that `depends_on "gcc@x"` makes just sure the dependency is installed, but the compiler selection is completely unrelated and can be something completely different. It's a little bit surprising: you define a compiler as a dependency, but a "random" unrelated compiler will be picked up.

It looks like the only way to control the selected compiler is to use `fails_with`.

Here is the combinations I tested:

```
fails_with gcc: "6"
depends_on "gcc@7"
```

-> will use system gcc 5

```
fails_with gcc: "5"
fails_with gcc: "6"
depends_on "gcc@7"
```

-> will use brewed gcc 9

```
fails_with gcc: "5"
fails_with gcc: "6"
fails_with gcc: "9"
depends_on "gcc@7"
```

-> will use brewed gcc 8

```
fails_with gcc: "5"
fails_with gcc: "6"
fails_with gcc: "8"
fails_with gcc: "9"
depends_on "gcc@7"
```

-> will use brewed gcc 7

```
fails_with gcc: "5"
fails_with gcc: "6"
fails_with gcc: "7"
fails_with gcc: "8"
fails_with gcc: "9"
depends_on "gcc@7"
```

-> will use brewed gcc (10)

So in the case I really just want to use gcc 7 for my build, I need to set:
```
fails_with gcc: "5"
fails_with gcc: "6"
fails_with gcc: "8"
fails_with gcc: "9"
depends_on "gcc@7"
```

It works, but it feels wrong to have to define all of these.

My dream syntax would be:

```
fails_with gcc: "6"
depends_on "gcc@7"
```

which would mean: all compilers below gcc 6 are incompatible, use brewed gcc 7 (or system gcc 7 I guess) only.

**Comment by danielnachun** [comment]
One important question that will affect how we resolve this is whether we want to always force the latest version of `gcc` if `gcc-5` (or `gcc-7` in the future) is not compatible.  

If we went down this path, it would be simpler from the perspective of dependencies (we wouldn't have to keep track of the minimum version of GCC that works for each formula), but would mean forcing the user to install brewed GCC even if they have a new enough system GCC.  

I am leaning in favor of this for a couple of reasons.  When we have fully relocatable binaries, users should rarely if ever be compiling from source.  Even in that scenario or for users who have `/home/linuxbrew/.linuxbrew` as their prefix, we still have to make sure the system GCC runtime libs are new enough to be compatible with whatever we built with in CI.

So for example if we used `gcc-8` to build in CI, we'd have to make sure that the user's system GCC was at least `gcc-8` - otherwise they would need brewed `gcc-8` installed to give them new enough runtime libs.

However, if we just always build anything that needs a newer GCC than `gcc-5` with `gcc-10`, then we can always use brewed `gcc-10` and be sure that we will have new enough runtime libs.  Although this is technically "wasteful" in the sense that a user who system GCC is `gcc-10` doesn't really need brewed `gcc-10`, for all the users who have an older GCC, this would be better than having 5 different versions GCC installed.

**Comment by iMichka** [comment]
We could always use gcc-10 for these cases, I agree, but it has two downsides:

- When we migrate from Ubuntu 16.04 to Ubuntu 18.04 for CI (which is the next big project I would like to start soon), we will have to downgrade all the formulae that depend on gcc-10 to system gcc-7 (if it builds with systems gcc-7).
- Some formulae will probably not build with gcc-10 because it's too new: we still will need to go back to gcc-7/8/9 for some of them.

**Comment by sjackman** [comment]
On Linuxbrew-core CI we have only two versions of GCC.
`/usr/bin/gcc-5`
`/home/linuxbrew/.linuxbrew/bin/gcc-10` (when installed)

`gcc-5` is the first priority.
`gcc-10` is the second priority.
No other versions of GCC matter for Linuxbrew-core CI, because they're not installed and not available for use, because formulae should not IMO depend on any other version of GCC. (We could consider very specific exceptions if absolutely necessary.)

If a formula fails to build with GCC 5, then specify `fails_with gcc: "5"`, and it will be built with GCC 10. If it has a runtime dependency on `gcc` (for `libstdc++.so.6` version ≥10 for example), then you also need `depends_on "gcc"`.

I don't believe that we need to worry about any other case of `fails_with gcc: "X"` other than 5, because no other compilers are available on Linuxbrew-core CI other than GCC versions 5 and 10.

Supporting users building from source with other versions of GCC is done on a best effort basis, but if building with a version of GCC other than 5 or 10 fails, the first thing the user should try is `brew install gcc`, and see if that resolves their issue. We can put that hint in the issue reporting template.

File: Formula/abseil.rb [hunk 1]
```diff
@@ -20,6 +20,8 @@ class Abseil < Formula
     fails_with gcc: "4"
     fails_with gcc: "5"
     fails_with gcc: "6"
+    fails_with gcc: "8"
+    fails_with gcc: "9"
     depends_on "gcc@7" => :build
```
Comment (sjackman) [inline Formula/abseil.rb:25]:
```suggestion
    fails_with gcc: "5"
```

**Review by sjackman — CHANGES_REQUESTED** [review]

**Comment by sjackman** [comment]
Abseil appears to work just fine with GCC 5. What's the motivation for not using GCC 5?

**Comment by danielnachun** [comment]
> Abseil appears to work just fine with GCC 5. What's the motivation for not using GCC 5?

It's possible that was done to deal with the issue of the brewed `gcc-5` being installed (incorrectly) in CI and its `libstdc++.so.6` being linked instead of the newer GCC.  In which case it may not be needed anymore.

**Comment by danielnachun** [comment]
> On Linuxbrew-core CI we have only two versions of GCC.
> `/usr/bin/gcc-5`
> `/home/linuxbrew/.linuxbrew/bin/gcc-10` (when installed)
> 
> `gcc-5` is the first priority.
> `gcc-10` is the second priority.
> No other versions of GCC matter for Linuxbrew-core CI, because they're not installed and not available for use, because formulae should not IMO depend on any other version of GCC. (We could consider very specific exceptions if absolutely necessary.)
> 
> If a formula fails to build with GCC 5, then specify `fails_with gcc: "5"`, and it will be built with GCC 10. If it has a runtime dependency on `gcc` (for `libstdc++.so.6` version ≥10 for example), then you also need `depends_on "gcc"`.
> 
> I don't believe that we need to worry about any other case of `fails_with gcc: "X"` other than 5, because no other compilers are available on Linuxbrew-core CI other than GCC versions 5 and 10.
> 
> Supporting users building from source with other versions of GCC is done on a best effort basis, but if building with a version of GCC other than 5 or 10 fails, the first thing the user should try is `brew install gcc`, and see if that resolves their issue. We can put that hint in the issue reporting template.

I like this as the approach.  We could just make exceptions if formula somehow need a version of gcc greater than 5 and less than 10 (I think this doesn't exist right now, and may never occur in practice).  

When I'm hopefully able to implement the binary patching for relocation solution, we won't have to worry much at all about users being able to build formulae with arbitrary versions of GCC.

The only gripe someone might have at that point  (and this is already true if they use the standard prefix `/home/linuxbrew/.linuxbrew`) could be that we are "needlessly" installing GCC as a runtime dependency when for some systems, the system GCC runtime libs might be new enough.  But I don't really see a reason why a runtime dependency on GCC is any different from a runtime dependency on any other software.  We don't try to see if the user has a new enough version of OpenJDK already installed, etc.

**Comment by sjackman** [comment]
I've opened two PRs:
- https://github.com/Homebrew/linuxbrew-core/pull/22485 `abseil, grpc: Build with GCC 5`
- https://github.com/Homebrew/linuxbrew-core/pull/22486 `bear: Style fixes`

**Comment by iMichka** [comment]
> On Linuxbrew-core CI we have only two versions of GCC.

On my computer I have more compilers. I might be on ubuntu 18.04 and have gcc 7. Or Ubuntu 20.04 with an even newer gcc.
We want to fix it for everybody, not only CI. I might want to compile something from source because I a maintainer. Or I might need to compile something from source because my bottle is not relocatable and I am building in a different prefix.

> gcc-5 is the first priority.
> gcc-10 is the second priority.

Not in my opinion. For me I think this is the priority list:
> gcc-5 is the first priority.
> gcc-7 is the second priority.

Because:
- this is how it always has been, so we do not change that behaviour for our users
- we when migrate to ubuntu 18.04 I do not need "downgrade" all these formulae from gcc-10 to gcc-7

**Comment by danielnachun** [comment]
So much for not having to worrying about formula which can only build with GCC versions greater than 5 and less than 10.  It looks like `protobuf` and `mysql` won't build with `gcc-5` because `libstdc++.so` is too old, but also can't build with `gcc-10` because the `cmath` GCC header file is too new.  

Perhaps we could track the Ubuntu LTS gcc versions?  So first `gcc-5`, then `gcc-7` (for 18.04 LTS), then `gcc-9` (for 20.04 LTS), and then `gcc-10`?

**Comment by iMichka** [comment]
>looks like protobuf and mysql won't build with gcc-5

Yes, found that out and wrote it here: https://github.com/Homebrew/linuxbrew-core/pull/22485#issuecomment-792529727

**Comment by sjackman** [comment]
>> gcc-5 is the first priority.
>> gcc-10 is the second priority.

> Not in my opinion. For me I think this is the priority list:

I'm describing the reality of Homebrew/brew here. That's the compiler preference scheme of Homebrew/brew.

**Comment by sjackman** [comment]
> It looks like protobuf and mysql won't build with gcc-5 because libstdc++.so is too old, but also can't build with gcc-10 because the cmath GCC header file is too new.

Does upstream or Debian have a patch for the GCC 10 issue?

**Comment by sjackman** [comment]
> It looks like protobuf and mysql won't build with gcc-5 because libstdc++.so is too old, but also can't build with gcc-10 because the cmath GCC header file is too new.

I just tested building `protobuf` with GCC 5, and it worked for me. See PR https://github.com/Homebrew/linuxbrew-core/pull/22498. I also tested `brew test apache-arrow bloaty grpc libphonenumber mitmproxy mosh nanopb-generator osm-pbf protobuf-c zbackup`, which succeeded. I did not test `mysql`.

**Comment by iMichka** [comment]
So finally everything just works by magic and we have no clue what happed. I mean, it's easier like this and if builds pass with system gcc-5, I'm happy. Let's remove the gcc@x versions when possible and have a look at what remains at the end.

> I'm describing the reality of Homebrew/brew here. That's the compiler preference scheme of Homebrew/brew.

I still think that design is wrong and should be changed (changing the reality to make it fit my reality). But if it impacts only a few formulae (or maybe even none) and nobody else than me complains, I'll give up on changing the whole logic because I don't have the energy to bikeshed this. @sjackman you won by working around my initial point :joy:

**Comment by sjackman** [comment]
> So finally everything just works by magic and we have no clue what happed.

If I were to take a guess, the impactful change may have been moving `gcc-5` up in the compiler priority list ahead of `gcc-10`, so that a more recent version of GCC wasn't pulled in opportunistically when not specifically required.

**Comment by sjackman** [comment]
> you won by working around my initial point 😂

I am the artful dodger. 😁

**Pull request closed without merging**